File: 01_two_pointers/02_3sum/3sum.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        answer = []
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j = i + 1
            k = len(nums) - 1
            while (i < j < k):
                sum = nums[i] + nums[j] + nums[k]
                if sum == 0:
                    answer.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while k > j and nums[k - 1] == nums[k]:
                        k -= 1
                    j += 1
                    k -= 1
                elif sum > 0:
                    k -= 1
                elif sum < 0:
                    j += 1
        return answer


if __name__ == "__main__":
    solution = Solution()
    nums = [-1, 0, 1, 2, -1, -4]
    answer = solution.threeSum(nums)
    print(answer)

File: NeetCode/01_two_pointers/02_3sum/3sum.cpp
#include <vector>
#include <algorithm>

using std::vector;

class Solution
{
  public:
    vector<vector<int> > threeSum(vector<int>& nums)
    {
        vector<vector<int> > solution;
        std::sort(nums.begin(), nums.end());

        size_t i = 0;
        while (i < nums.size() - 2)
        {
            if (i > 0 && nums[i - 1] == nums[i]) // skip checked duplicates
            {
                ++i;
                continue;
            }
            size_t j = i + 1;
            size_t k = nums.size() - 1;
            while (i < j && j < k)
            {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == 0)
                {
                    solution.push_back(vector<int>());
                    solution.back().push_back(nums[i]);
                    solution.back().push_back(nums[j]);
                    solution.back().push_back(nums[k]);
                    while (j < k && nums[j + 1] == nums[j]) // skip duplicates
                        ++j;
                    while (k > j && nums[k - 1] == nums[k]) // skip duplicates
                        --k;
                    ++j;
                    --k;
                }
                else if (sum > 0)
                    --k;
                else if (sum < 0)
                    ++j;
            }
            ++i;
        }
        return solution;
    }
};

int main()
{
    Solution solution;

    vector<int> vec;
    vec.push_back(-1);
    vec.push_back(0);
    vec.push_back(1);
    vec.push_back(2);
    vec.push_back(-1);
    vec.push_back(-4);
    // [-1,0,1,2,-1,-4]
    vec.push_back(0);
    vec.push_back(0);
    vec.push_back(0);
    vector<vector<int> > answer = solution.threeSum(vec);
}
